src/WPGraphQL.php

```diff
@@ -422,50 +424,127 @@ public function init_admin() {
-			$wp_taxonomies['category']->graphql_single_name = 'category';
-			$wp_taxonomies['category']->graphql_plural_name = 'categories';
+		return $args;
+	}
+
+	/**
+	 * This sets the post type /taxonomy GraphQL properties.
+	 *
+	 * @param array  $args      Array of arguments for registering the type.
+	 * @param string $wp_type   Post type / Taxonomy type key.
+	 *
+	 * @return array
+	 */
+	public static function set_wp_type_args( $args, $wp_type ) {
+
+		// Bail early if the post type is hidden from the WPGraphQL schema.
+		if ( empty( $args['show_in_graphql'] ) ) {
+			return $args;
 		}
 
-		// Adds GraphQL support for tags.
-		if ( isset( $wp_taxonomies['post_tag'] ) ) {
-			$wp_taxonomies['post_tag']->show_in_graphql     = true;
-			$wp_taxonomies['post_tag']->graphql_single_name = 'tag';
-			$wp_taxonomies['post_tag']->graphql_plural_name = 'tags';
+		$defaults = [
+			// The "kind" of GraphQL type to register. Can be `interface`, `object`, or `union`.
+			'graphql_kind'                     => 'object',
+			// The callback used to resolve the type. Only used if `graphql_kind` is an `interface` or `union`.
+			'graphql_resolve_type'             => null,
+			// An array of custom interfaces the type should implement.
+			'graphql_interfaces'               => [],
+			// An array of default interfaces the type should exclude.
+			'graphql_exclude_interfaces'       => [],
+			// An array of custom connections the type should implement.
+			'graphql_connections'              => [],
+			// An array of default connection field names the type should exclude.
+			'graphql_exclude_connections'      => [],
+			// An array of possible type the union can resolve to. Only used if `graphql_kind` is a `union`.
+			'graphql_union_types'              => [],
+			// Whether to register default connections to the schema.
+			'graphql_register_root_field'      => true,
+			'graphql_register_root_connection' => true,
+		];
+
+		$args = wp_parse_args( $args, $defaults );
+
+		// Ensure a valid type resolver is set for interface and union types.
+		if ( ! is_callable( $args['graphql_resolve_type'] ) && ( 'interface' === $args['graphql_kind'] || 'union' === $args['graphql_kind'] ) ) {
+			throw new Exception(
+				sprintf(
+					__( '%1$s is registered as a GraphQL %2$s, but has no way to resolve the type. Ensure $args[\'graphql_resolve_type\'] is a valid callback function.', 'wp-graphql' ),
+					$wp_type,
+					$args['graphql_kind']
+				)
+			);
 		}
 
-		// Adds GraphQL support for post formats.
-		if ( isset( $wp_taxonomies['post_format'] ) ) {
-			$wp_taxonomies['post_format']->show_in_graphql     = true;
-			$wp_taxonomies['post_format']->graphql_single_name = 'postFormat';
-			$wp_taxonomies['post_format']->graphql_plural_name = 'postFormats';
+		// Ensure union types have a valid list of possible types.
+		if ( empty( $args['graphql_union_types'] ) && 'union' === $args['graphql_kind'] ) {
+			throw new Exception(
```

Comment:
I think this might be too aggressive. I'm seeing fatal errors when I register a post type like so, then load the admin page: 

```php
add_action( 'init', function() {

	register_post_type( 'testType', [
		'show_in_graphql'     => true,
		'graphql_single_name' => 'TestType',
		'graphql_plural_name' => 'TestTypes',
		'graphql_kind'        => 'union',
	] );

} );
```

I think we should be more graceful with this, and either show an error in GraphQL requests, or skip the registration of the invalid type and show a Debug message. 

In any case, I think it should be limited to GraphQL environment rather than throwing an Exception on every WordPress load. 

![CleanShot 2022-09-08 at 13 05 32](https://user-images.githubusercontent.com/1260765/189205589-779b1fe1-7ca6-4675-bc04-534e983f1f5b.png)

Reply:
merged: https://github.com/wp-graphql/wp-graphql/pull/2399/commits/263be0cc1aa5fc2296b99dce373d5924cb94ad9f